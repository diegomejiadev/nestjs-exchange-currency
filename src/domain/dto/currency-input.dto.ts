import {
  IsNotEmpty,
  IsNumber,
  IsPositive,
  IsString,
  Min,
} from 'class-validator';

export class CurrencyInputDto {
  @Min(0, { message: 'Debe ingresar un valor mayor o igual a 0' })
  @IsNumber(
    { allowInfinity: false, allowNaN: false },
    { message: 'El monto ingresado debe ser num√©rico' },
  )
  @IsNotEmpty({ message: 'Debe ingresar el monto' })
  amount: number;

  @IsString({ message: 'Debe ingresar un texto en la moneda de origen' })
  @IsNotEmpty({ message: 'Debe ingresar una moneda de origen' })
  originCurrency: string;

  @IsString({ message: 'Debe ingresar un texto en la moneda de destino' })
  @IsNotEmpty({ message: 'Debe ingresar una moneda de destino' })
  destinyCurrency: string;
}
