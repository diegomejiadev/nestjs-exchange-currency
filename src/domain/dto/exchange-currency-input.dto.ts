import { ApiProperty } from '@nestjs/swagger';
import { IsNotEmpty, IsNumber, IsString, Min } from 'class-validator';

export class ExchangeCurrencyInputDto {
  @ApiProperty({
    description: 'Valor inicial que va a ser convertido',
    default: 10,
    type: Number,
  })
  @Min(0, { message: 'Debe ingresar un valor mayor o igual a 0' })
  @IsNumber(
    { allowInfinity: false, allowNaN: false },
    { message: 'El monto ingresado debe ser num√©rico' },
  )
  @IsNotEmpty({ message: 'Debe ingresar el monto' })
  amount: number;

  @ApiProperty({
    description: 'Tipo de moneda de origen',
    default: 'USD',
  })
  @IsString({ message: 'Debe ingresar un texto en la moneda de origen' })
  @IsNotEmpty({ message: 'Debe ingresar una moneda de origen' })
  originCurrency: string;

  @ApiProperty({
    description: 'Tipo de moneda de destino',
    default: 'PEN',
  })
  @IsString({ message: 'Debe ingresar un texto en la moneda de destino' })
  @IsNotEmpty({ message: 'Debe ingresar una moneda de destino' })
  destinyCurrency: string;
}
