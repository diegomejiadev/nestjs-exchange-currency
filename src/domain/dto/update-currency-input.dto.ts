import { ApiProperty } from '@nestjs/swagger';
import {
  IsNotEmpty,
  IsString,
  ValidationArguments,
  ValidationOptions,
  isObject,
  registerDecorator,
} from 'class-validator';

const IsRecord = (validationOptions?: ValidationOptions) => {
  return function (object: unknown, propertyName: string) {
    registerDecorator({
      name: 'IsRecord',
      target: object.constructor,
      propertyName: propertyName,
      constraints: [],
      options: {
        message: 'Wrong object format',
        ...validationOptions,
      },
      validator: {
        validate(value: unknown, args: ValidationArguments) {
          if (!isObject(value)) return false;
          if (Object.keys(value).length === 0) return true;

          const keys = Object.keys(value);

          return keys.every((key) => {
            if (typeof key !== 'string') return false;
            if (typeof value[key] !== 'number') return false;

            return true;
          });
        },
      },
    });
  };
};

export class UpdateCurrencyInputDto {
  @ApiProperty({
    description: 'CÃ³digo del tipo de moneda a actualizar',
    default: 'PEN',
    type: String,
  })
  @IsString({ message: 'Debe ingresar un texto del tipo de moneda' })
  @IsNotEmpty({ message: 'Debe ingresar un tipo de moneda' })
  code: string;

  @ApiProperty({
    description: 'Monedas a las que se les va a actualizar el tipo de cambio',
    default: {
      USD: 3.7,
      EUR: 4,
    },
  })
  @IsRecord({
    message:
      'Los valores en los tipos de moneda destino a actualizar debe ser { "COD": 123, "CAD": 456 }',
  })
  @IsNotEmpty({
    message: 'Debe ingresar los valores a actualizar en los tipos de moneda',
  })
  data: Record<string, number>;
}
