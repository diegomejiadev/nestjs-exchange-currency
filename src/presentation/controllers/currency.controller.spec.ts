import { Test, TestingModule } from '@nestjs/testing';
import { CurrencyService } from '../services/currency.service';
import { CurrencyController } from './currency.controller';
import { ExchangeCurrencyInputDto } from 'src/domain/dto/exchange-currency-input.dto';
import { ExchangeCurrencyEntity } from 'src/domain/entities/exchange-currency.entity';
import { MockCurrencyService } from './mock-currency.service';

describe('CurrencyController', () => {
  let controller: CurrencyController;
  let service: CurrencyService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      controllers: [CurrencyController],
      providers: [CurrencyService],
    })
      .overrideProvider(CurrencyService)
      .useClass(MockCurrencyService)
      .compile();

    controller = module.get<CurrencyController>(CurrencyController);
    service = module.get<CurrencyService>(CurrencyService);
  });

  it('should be defined', () => {
    expect(controller).toBeDefined();
  });

  it('should exchange a currency', async () => {
    const body: ExchangeCurrencyInputDto = {
      amount: 5,
      destinyCurrency: 'PEN',
      originCurrency: 'USD',
    };

    const result: ExchangeCurrencyEntity = {
      amount: expect.any(Number),
      baseAmount: expect.any(Number),
      destinyCurrency: 'PEN',
      exchangeType: 'USD -> PEN',
      originCurrency: 'USD',
      lastUpdated: expect.any(String),
    };

    expect(await controller.exchangeCurrency(body)).toEqual(result);

    const exchangeCurrencySpy = jest.spyOn(service, 'exchangeCurrency');
    await service.exchangeCurrency(body);

    expect(exchangeCurrencySpy).toHaveBeenCalledWith(body);
  });

  it('should load all currencies', async () => {
    expect(await controller.loadAllCurrencies()).toBe(true);
  });

  it('list all currencies', async () => {
    const result = await controller.listAllCurrencies();

    result.forEach((resultItem) => {
      expect(resultItem).toMatchSnapshot({
        code: expect.any(String),
        updated_at: expect.any(String),
      });
    });
  });
});
