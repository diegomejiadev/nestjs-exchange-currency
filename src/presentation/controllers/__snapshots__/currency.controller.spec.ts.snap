// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyController list all currencies 1`] = `
{
  "code": Any<String>,
  "data": {
    "EUR": {
      "code": "EUR",
      "value": 0.25,
    },
    "USD": {
      "code": "USD",
      "value": 0.27,
    },
  },
  "updated_at": Any<String>,
}
`;

exports[`CurrencyController list all currencies 2`] = `
{
  "code": Any<String>,
  "data": {
    "EUR": {
      "code": "EUR",
      "value": 0.91,
    },
    "PEN": {
      "code": "PEN",
      "value": 3.7,
    },
  },
  "updated_at": Any<String>,
}
`;
